runtime/lua/vim/shared.lua

```diff
@@ -983,4 +983,22 @@ do
   end
 end
 
+--- @generic T
+--- @param root string
+--- @param _mod T
+--- @return T
+function vim._deferload_submodules(root, _mod)
+  return setmetatable({}, {
+    __index = function(t, k)
+      local ok, ret = pcall(require, string.format('%s.%s', root, k))
```

Comment:
```suggestion
    ---@param t table<string, any>
    ---@param k string
    __index = function(t, k)
      if not _mod[k] then
        return nil
      end
      local modname = string.format('%s.%s', root, k)
      ---@type boolean, any
      local ok, ret = pcall(require, modname)
```


* `vim.treesitter.submodule_that_does_not_exist` should return nil (without errors)

* `vim.treesitter.submodule_to_defer_load` should throw errors when an error would happen while `require`-ing that module.

In order to make this possible, we'll need to change the "deferload" declaration as:

```diff
 local M = vim._deferload_submodules('vim.treesitter', {
-  _fold = ..., --- @module 'vim.treesitter._fold'
+  _fold = { ... }, --- @module 'vim.treesitter._fold'
```

because `...` evaluates to `nil` in Lua. The table `_mod` is currently not used, but can be used to determine which submodule is registered to be lazy-loaded.

Reply:
Actually with this we can completely remove `pcall(...)` with just simple `require(...)`. Stacktrace might be different (i.e. whether to include additional frame on `__index`), but I haven't tested. Doesn't matter much though.

Reply:
That's pretty insane tbh, this is getting out of hand lol

Reply:
What do you mean by "should"? As in that is the behavior on master and that we should emulate it?

Reply:
It should work as expected now. I don't think wrapping it around a table is necessary.

Reply:
I'm not sure why it's obvious `vim.treesitter.submodule_that_does_not_exist` should return nil without errors. It's not really a table after all; it's a require.

Reply:
The current behaviour is that it returns nil, and I see no reason we should change that.

Reply:
I don't think we need to encase the modules in tables to do this. Thanks anyway.

Reply:
> I don't think we need to encase the modules in tables to do this.

Hmm you're right. When require-ing the module, `...` evaluates to the name of the module (e.g. the `"vim.treesitter"` string), *not `nil`*. (Maybe I did something wrong when I was testing this out yesterday...!?) Sorry about making a confusion here.